### PR TITLE
fix(sglang): fix sglang oom (Qwen3.5-35B-A3B)

### PR DESCRIPTION
**Pull request by MengsD**
## Description

When running Qwen3.5-35B-A3B (Hybrid MoE + Mamba, ~63.4 GB total weights) with NCCL-based weight sync from Trainer to SGLang Explorer, SGLang allocates receive buffers
 for all tensors simultaneously, causing OOM on H20 GPUs (95 GiB each, ~50 GiB available after model loading).

 A secondary OOM occurs on Hybrid Mamba models because SGLang pre-allocates Mamba state memory (mamba_full_memory_ratio=0.9), consuming nearly all remaining GPU memory —
  leaving no room even for batched receive buffers.

 ## Fix

  Commit 1 — Batched weight sync:
  - Partition state_dict_meta into configurable batches (default 4 GB each, controlled by InferenceModelConfig.weight_sync_buffer_size)
  - Call update_weights_from_distributed once per batch instead of once for all tensors
  - SGLang allocates/releases buffers per batch, keeping peak memory bounded
  
  Commit 2 — Flush cache before sync:
  - Call SGLang's /flush_cache endpoint before starting NCCL weight sync
  - Releases KV + Mamba cache (~73 GiB across 8 GPUs), making room for receive buffers
  - Cache is automatically repopulated on next inference request
  
 ## Files Changed

  - trinity/common/models/sglang_model.py — batching logic, flush_cache() client method, sync flow update
  - trinity/common/config.py — add weight_sync_buffer_size field to InferenceModelConfig

## What to Verify

  1. Weight sync completes without OOM — look for log lines:
  Flushed KV/Mamba cache before NCCL weight sync
  NCCL weight sync: XXX tensors in 21 batches (buffer_size=4.0 GB)
  Synchronization finished.
  2. Rollout succeeds after sync — model generates valid responses after weights are updated
  3. Training step proceeds — trainer consumes the rollout experiences and completes a step
  Verified
  
  
example.yaml 👇
  
 

> project: Trinity-RFT-example
name: trinity-dapo
checkpoint_root_dir: ${oc.env:TRINITY_CHECKPOINT_ROOT_DIR,./checkpoints}
model:
  model_path: ${oc.env:TRINITY_MODEL_PATH} # Suggest using larger model on this dataset
  max_response_tokens: 256
  max_model_len: 1024
algorithm:
  enable_router_replay: true
  algorithm_type: grpo
  repeat_times: 4
  optimizer:
    lr: 1e-6
    lr_warmup_steps: 20
  policy_loss_fn_args:
    clip_range_low: 0.2
    clip_range_high: 0.28
cluster:
  node_num: 4
  gpu_per_node: 8
buffer:
  total_steps: 2
  batch_size: 8
  explorer_input:
    taskset:
      name: dapo-math
      storage_type: file
      path: /nas/xiqing/datasets/trinity/dapo-math-17k
      subset_name: all
      format:
        prompt_key: 'prompt'
        response_key: 'solution'
        system_prompt: 'Solve the following math problem step by step. The last line of your response should be of the form Answer: $Answer (without quotes) where $Answer is the answer to the problem.'
      rollout_args:
        temperature: 1.0
      workflow_args:
        use_base: true
      reward_fn_args:
        enable_overlong_penalty: true
        penalty_factor: 1.0
        max_response_length: 8192
        cache_length: 256
    eval_tasksets:
    - name: AIME2024
      storage_type: file
      path: /nas/xiqing/datasets/trinity/aime24
      split: 'train'
      repeat_times: 2
      format:
        prompt_key: 'question'
        response_key: 'answer'
        system_prompt: 'Solve the following math problem step by step. The last line of your response should be of the form Answer: $Answer (without quotes) where $Answer is the answer to the problem.'
      rollout_args:
        temperature: 1.0
        top_p: 1
    default_workflow_type: 'math_boxed_workflow'
    default_reward_fn_type: 'math_dapo_reward'
  trainer_input:
    experience_buffer:
      name: math_buffer
      storage_type: queue
explorer:
  eval_interval: 20
  runner_per_model: 4
  rollout_model:
    engine_type: sglang
    engine_num: 4
    tensor_parallel_size: 4
    enforce_eager: true
    enable_prefix_caching: true
    enable_chunked_prefill: true
    gpu_memory_utilization: 0.4
    dtype: bfloat16
    seed: 42
synchronizer:
  sync_method: 'nccl'
  sync_interval: 16
  sync_timeout: 1200
trainer:
  save_interval: 100
  grad_clip: 1.0
  use_dynamic_bsz: true
  max_token_len_per_gpu: 256
  trainer_strategy: megatron
  trainer_config:
    actor_rollout_ref:
      actor:
        megatron:
          tensor_model_parallel_size: 4
          pipeline_model_parallel_size: 1
          expert_model_parallel_size: 8
          expert_tensor_parallel_size: 1
          context_parallel_size: 1
          sequence_parallel: true
          use_dynamic_bsz: true
          use_remove_padding: true
          use_distributed_optimizer: true
          override_transformer_config:
          recompute_granularity: full
          recompute_method: uniform
          recompute_num_layers: 1

## Checklist

Please check the following items before code is ready to be reviewed.

- [ ]  Code has passed all tests
- [ ]  Docstrings have been added/updated in Google Style
- [ ]  Documentation has been updated
- [ ]  Code is ready for review

### CODE REVIEW

**Comment by MengsD** [comment]
@claude can you review the code ?

**Review by pan-x-c — APPROVED** [review]
